packages/shared/src/telemetry/collector.ts

```diff
@@ -0,0 +1,300 @@
+
+    // We will try to grab the SDK data lazily when an event is triggered, so it should always be defined once the event is sent.
+    this.#metadata.sdk = options.sdk!;
+    this.#metadata.sdkVersion = options.sdkVersion!;
+
+    this.#metadata.publishableKey = options.publishableKey ?? '';
+
+    const parsedKey = parsePublishableKey(options.publishableKey);
+    if (parsedKey) {
+      this.#metadata.instanceType = parsedKey.instanceType;
+    }
+
+    if (options.secretKey) {
+      // Only send the first 16 characters of the secret key to to avoid sending the full key. We can still query against the partial key.
+      this.#metadata.secretKey = options.secretKey.substring(0, 16);
+    }
+  }
+
+  get isEnabled(): boolean {
+    if (this.#metadata.instanceType !== 'development') {
+      return false;
+    }
+
+    // In browser or client environments, we most likely pass the disabled option to the collector, but in environments
+    // where environment variables are available we also check for `CLERK_TELEMETRY_DISABLED`.
+    if (this.#config.disabled || (typeof process !== 'undefined' && isTruthy(process.env.CLERK_TELEMETRY_DISABLED))) {
+      return false;
+    }
+
+    return true;
+  }
+
+  get isDebug(): boolean {
+    return this.#config.debug || (typeof process !== 'undefined' && isTruthy(process.env.CLERK_TELEMETRY_DEBUG));
+  }
+
+  record(event: Pick<TelemetryEvent, 'event' | 'payload'>): void;
+  record(event: TelemetryEvent['event'], payload: TelemetryEvent['payload']): void;
+  record(
+    event: TelemetryEvent['event'] | Pick<TelemetryEvent, 'event' | 'payload'>,
+    payload?: TelemetryEvent['payload'],
+  ): void {
+    const preparedPayload =
+      typeof event === 'string'
+        ? this.#preparePayload(event, payload ?? {})
+        : this.#preparePayload(event.event, event.payload);
+
+    this.#logEvent(preparedPayload.event, preparedPayload);
+
+    if (!this.#shouldRecord()) {
+      return;
+    }
+
+    this.#buffer.push(preparedPayload);
+
+    this.#scheduleFlush();
+  }
+
+  #shouldRecord(): boolean {
+    return this.isEnabled && !this.isDebug && Math.random() <= this.#config.samplingRate;
+  }
+
+  #scheduleFlush(): void {
+    // On the server, we want to flush immediately as we have less guarantees about the lifecycle of the process
+    if (typeof window === 'undefined') {
+      this.#flush();
+      return;
+    }
+
+    const isBufferFull = this.#buffer.length >= this.#config.maxBufferSize;
+    if (isBufferFull) {
+      // If the buffer is full, flush immediately to make sure we minimize the chance of event loss.
+      // Cancel any pending flushes as we're going to flush immediately
+      if (this.#pendingFlush) {
+        const cancel = typeof cancelIdleCallback !== 'undefined' ? cancelIdleCallback : clearTimeout;
+        cancel(this.#pendingFlush);
+      }
+      this.#flush();
+      return;
+    }
+
+    // If we have a pending flush, do nothing
+    if (this.#pendingFlush) return;
```

Comment:
How come eslint is not complaining about the missing braces here? 🤔

Reply:
Not sure, do we have that rule enabled? 💭

Reply:
Interesting, let me check when we and why we removed this one, probably by accident.

Reply:
Marking this as resolved as it is unrelated to this PR. I posted more on slack